pan/functions.pan

```diff
@@ -1,270 +1,290 @@
-        error("domain_from_object: invalid default domain ("+ARGV[0]+")");
+    if (ARGC != 1) {
+        error("usage: domain_from_object(default_domain) requires default domain as argument");
     };
 
-    # Check if the name matches.  We can include the 'profile_' in
-    # the host part if it exists.  We only need the domain here.
-    m = matches(OBJECT,"([^\\.]+)(\\.(.*))?");
-    size = length(m);
-    if (size >= 2) {
-        if (size >= 4) {
-            if (is_fqdn(m[3])) {
-                return(m[3]);
-            } else {
-                error("domain_from_object: invalid domain name ("+m[3]+")");
-            };
-        } else {
-            return(ARGV[0]);
-        };
-    } else {
-        error("domain_from_object: can't match domain name from object");
-        };
+    host_domain = host_domain_from_object(ARGV[0]);
+    host_domain[1];
 };
 
 
-########################################################################
-##=
-## @function full_hostname_from_object
-## @# extracts the full hostname from the object template name.
-##+If the domain information is not in the name, the default is used.
-##+This assumes that your machine profile has the form
-##+profile_*.  If it does not, an error is thrown.
-## @syntax domain:string
-## @param:domain default domain name
-## @example
-##+# "/data" will contain "myhost.example.net"
-##+object template profile_myhost;
-##+"/data" = hostname_from_object("example.net");
-##=
-############################################################
+@documentation{
+    extracts the full hostname from the object template name.
+    If the domain information is not in the name, the default is used.
+
+    @syntax domain:string
+    @param:domain default domain name
+
+    Example:
+    object template myhost;
+    "/data" = hostname_from_object("example.net");
+
+    "/data" will contain "myhost.example.net"
+}
 function full_hostname_from_object = {
     # Check cardinality; leave detailed error checking to called functions.
-    if (ARGC != 1) error("usage: full_hostname_from_object(default_domain)");
+    if (ARGC != 1) error("usage: full_hostname_from_object(default_domain) requires default_domian as argument");
 
-    return(hostname_from_object()+"."+domain_from_object(ARGV[0]));
+    host_domain = host_domain_from_object(ARGV[0]);
+    format("%s.%s", host_domain[0], host_domain[1]);
 };
 
+@documentation{
+    Test if (all elements of) the first argument are
+    any of the remaining arguments.
+    (If the first argument is a list, all its elements are tested).
+
+    Example:
+    is_choice_of(1, 1, 2, 3) returns true
+    is_choice_of(1, 2, 3) returns false (a string is not a long)
+
+    is_choice_of(list("b", "c"), "a", "b", "c") returns true
+    is_choice_of(list("a", "d"), "a", "b", "c") returns false ("d" is not an argument)
+
+    The types are not checked and are assumed to be comparable.
+        is_choice_of("1", 1, 2, 3) will give a panc error (string vs long),
```

Comment:
Not very good... For me this is something to avoid. The function should return an understandable message in case of type mismatch. Personnally I'll enforce type checking, this wil add a real value to `is_choice_of()`.
* First check that all valid values passed have the same type
* Second check that any value to check if of this type